copy/copy.go

```diff
@@ -1098,6 +1098,18 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, to
 					Artifact: srcInfo,
 				}
 			}
+
+			// If we know the blob that we just reused is compressed (or not), note how it's compressed (or not).
```

Comment:
Assuming this does the same thing as `dockerImageDest`: It’s just _ugly_ to compute the same thing twice. Options:
- Remove this code from `dockerImageDest`, and leave it only in the generic path. Conceptually a better match, OTOH a little performance impact due to locking the BIC over and over (and we can _almost_ remove `CandidateLocations2` again, except for the part where it drops `UnknownCompression` candidates)
- Have the blob cache deal with this, consistently with `dockerImageDest`, one way or another (make BIC2 public, move blob cache to this repo, …)
- Keep both, but only use the generic path if `TryReusingBlob` has not set `CompressionOperation`/`CompressionAlgorithm`
- Something else?

Reply:
I agree that we need to keep `CandidateLocations2` to filter out `UnknownCompression` candidates.  We also need to be able to look up compression info even if the destination repository isn't a location we've seen the blob before, so we need to be able to look up compression info independent of transport scopes, which `CandidateLocations` can't do, so I don't think we can avoid something like `DigestCompressorName()`.

Moving the logic that sets the `CompressionOperation` and `CompressionAlgorithm` mostly out of `dockerImageDest.TryReusingBlob()` into `copy` also works, but as you note below, `TryReusingBlob()` still needs to learn to only try its initial `HasBlob()` check if it knows how the blob was compressed, so that it won't return a `BlobInfo` that specifies a digest whose compression type we don't know yet before trying locations from the blob info cache.

Treating reuse of a blob for which we don't know the compression type as an error breaks buildah's commit routine, which returns diffIDs from its `LayerInfosForCopy()` to avoid needing to recompress anything, and it knows that the containers-storage `TryReusingBlob()` will succeed anyway.  Working around that by having the containers-storage `PutBlob()` add a compression note to the blob info cache works around that, but I'm not sure the combination is robust enough for other users of the library.